### DISCUSSION TITLE
High content sitemap

### DISCUSSION THREAD

**Discussion by mvneobux** [spatie/laravel-sitemap#469] [General]
From what I read in the documentation I didn't see an example that fits my scenario.

I have 11 tables (models) and each one with about 320k posts and I need to generate sitemap for each one of them. But I need to limit to 46 thousand lines in the sitemap. But from what I read it is only possible to limit the number of lines (maxTagsPerSitemap) if I configure it to crawl my site. Am I right ?

or can I easily implement the sitemap in each table and set the tag limit?

I would like to do something like this:

`
 SitemapGenerator::create(aa::all())
    ->maxTagsPerSitemap(10000)
    ->writeToFile(public_path('sitemap.xml'));
`

How can I do this ?

**Comment by anugupta** [comment]
So, the way I'm doing this (for a model with > 200k records is):

- create sitemaps for each chunk 
- add these sitemaps to a sitemap index

In the model implement the `Sitemapable` interface:

```

class Performance extends Model implements Sitemapable

{
    public function toSitemapTag(  ): Url | string | array
    {
        return route( 'performance.show', $this );
    }
    
}
```


And then have a service or helper or controller or whatever to generate the sitemaps and a sitemap index:

```
    public function generateSitemap()
    {

        $counter = 0;
        $sitemapIndex = SitemapIndex::create();

        Performance::chunk( 40000, function ( $chunk ) use ( &$counter, $sitemapIndex ) {

            $counter++;

            Sitemap::create()
                   ->add( $chunk )
                   ->writeToFile( public_path( 'sitemap-performances-' . $counter . '.xml' ) );

            $sitemapIndex->add( '/sitemap-performances-' . $counter . '.xml'  );

        } );

        $sitemapIndex->writeToFile( public_path( 'sitemap_index.xml' ) );

    }
```

Runs reasonably well, on my local dev environment, it creates the sitemap files for 200k records in under 4 minutes (would be much quicker but I have some scopes on the model that query other relationships to detemine validity - on a simpler model with around 30k records, it completes in a few seconds).

**Reply by mvneobux** [reply]
Thanks a lot for the suggestion. In mine it is only generating the sitemapindex. How does $chunk know what a URL or a date is?

I don't know if I'm doing it right, here's how I'm doing it:

my route/web.php
`
<?php

use Illuminate\Support\Facades\Route;

use Spatie\Sitemap\Contracts\Sitemapable;
use Spatie\Sitemap\Tags\Url;
use Spatie\Sitemap\Sitemap;
use Spatie\Sitemap\SitemapGenerator;
use Spatie\Sitemap\SitemapIndex;

use App\Models\palavras_br;


set_time_limit(0);
ini_set('memory_limit', '-1');
ini_set("max_execution_time", 0);

Route::get('/mapteste', function () {


        $counter = 0;
        $sitemapIndex = SitemapIndex::create();

        palavras_br::chunk( 40000, function ( $chunk ) use ( $counter, $sitemapIndex ) {
            
            
            $counter++;

            Sitemap::create()
                   ->add( $chunk )
                   ->writeToFile('palavras-' . $counter . '.xml');

            $sitemapIndex->add( 'sitemaps/teste/palavras-' . $counter . '.xml'  );

        } );

        $sitemapIndex->writeToFile('sitemaps/teste/sitemap_index.xml');
   

})->name('mapteste.show');
`

my model: 

`
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Spatie\Sitemap\Contracts\Sitemapable;
use Spatie\Sitemap\Tags\Url;

class palavras_br extends Model
{
    use HasFactory;
    
    protected $fillable = [
        'palavras',
    ];

    public $timestamps = false;
    
    protected $table = 'palavras_br';
    
    public function toSitemapTag(  ): Url | string | array
    {
        return route( 'mapteste.show', $this );
    }

}
`

Why doesn't it work? Only the sitemap index generates correctly

**Comment by anugupta** [comment]
Hard to tell what's wrong from that partial code fragment -  can you post it all and format it as a code block?

Have you implemented`toSitemapTag` in your model (first part of my answer) - that is what provides the URL to the `add` method.
